### PR TITLE
Make helm more customizable for the end user

### PR DESCRIPTION
**Pull request by clementrey-dev**
Issue #, if available:

Dear team

Description of changes:

The objective of this PR is to make the helm chart more customizable for the end user.
* customization of the image
* customization of the annotations for the service account
* customization of the placement constraint

The changes can described as follow:
* Externalize karpenter configuration by modifying the existing helm chart
  * modification of the kubernetes deployment so it integrates helm tag for templating
  * modification of the  service account definition so it integrates helm tag for templating
  * addition of a file `values.yaml`  specifying the karpenter configuration that can be externalized
  * modification of the `Makefile` to make all the targets working with helm
  * adding a test to verify the linting of the helm chart
 
* Simplifying the chart releasing process
  * deletion of a folder and files (under `releases/aws`) that was use in the previous process for building the helm chart
  * deletion of the associated targets in the `Makefile`.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

Hope this helps

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -41,22 +41,26 @@ licenses: ## Verifies dependency licenses and requires GITHUB_TOKEN to be set
...
-	ko delete -f config
+	helm template karpenter ./config/ |  ko delete -f -
 
 codegen: ## Generate code. Must be run if changes are made to ./pkg/apis/...
 	./hack/codegen.sh
 
 publish: ## Generate release manifests and publish a versioned container image.
 	@aws ecr-public get-login-password --region us-east-1 | docker login --username AWS --password-stdin $(RELEASE_REPO)
-	$(WITH_RELEASE_REPO) $(WITH_GOFLAGS) ko resolve -B -t $(RELEASE_VERSION) --platform all -f config > $(RELEASE_MANIFEST)
+	@mkdir -p $(RELEASE_CHART)/karpenter/templates
```
Comment (ellistarn) [inline Makefile:54]: I'd remove this line. We can assume the folder is already there.

File: Makefile [hunk 2]
```diff
@@ -1,6 +1,6 @@
 RELEASE_REPO ?= public.ecr.aws/karpenter
 RELEASE_VERSION ?= $(shell git describe --tags --always)
-RELEASE_MANIFEST = releases/$(CLOUD_PROVIDER)/manifest.yaml
+RELEASE_CHART = charts
```
Comment (ellistarn) [inline Makefile:3]: Any reason to make this dynamic? What is the usecase for changing this variable?

File: Makefile [hunk 3]
```diff
@@ -41,22 +41,26 @@ licenses: ## Verifies dependency licenses and requires GITHUB_TOKEN to be set
...
+	@cp config/templates/* $(RELEASE_CHART)/karpenter/templates
+	$(WITH_RELEASE_REPO) $(WITH_GOFLAGS) ko resolve -B -t $(RELEASE_VERSION) --platform all -f config/values.yaml > $(RELEASE_CHART)/karpenter/values.yaml
+	yq e -i '.version = "$(RELEASE_VERSION)"' ./$(RELEASE_CHART)/karpenter/Chart.yaml
 
-helm: ## Generate Helm Chart
-	cp $(RELEASE_MANIFEST) charts/karpenter/templates
-	yq e -i '.version = "$(RELEASE_VERSION)"' ./charts/karpenter/Chart.yaml
-	cd charts; helm package karpenter; helm repo index .
+verifiy-helm:
+	helm lint $(RELEASE_CHART)/karpenter
```
Comment (ellistarn) [inline Makefile:60]: This is awesome. Do you mind adding this step to the `helm` target instead of making a new step?

File: config/values.yaml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  defaultImage: ko://github.com/awslabs/karpenter/cmd/controller
+# fill the variable below if you want to override the default image
+#  image:
+#  tag:
+
+nodeSelector: {}
+
+tolerations: []
+
+affinity:
```
Comment (ellistarn) [inline config/values.yaml:15]: Karpenter builds multiple images for different architectures, which are automatically detected from the image manifest. I think we can get away without these node affinities.

File: config/Chart.yaml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+apiVersion: v2
+name: karpenter
+description: A Helm chart for https://github.com/awslabs/karpenter/.
+type: application
+version: 0.1
```
Comment (ellistarn) [inline config/Chart.yaml:5]: Can you lock this version to the current repo's version v0.2.3?

**Review by ellistarn — COMMENTED** [review]
Awesome PR, thank you so much! One thing it's missing is support for `make codegen`. Our codegen needs to drop generated manifests into the new location.

Also -- I think we can now delete the folder at /charts/karpenter in favor of /config.

Another option (which I think is cleaner), is to deprecate the config folder entirely, and instead do it all under /charts like: 

```
/charts/karpenter/Chart.yaml
/charts/karpenter/templates/...
/charts/karpenter-v0.2.3.tgz
```

**Comment by ellistarn** [comment]
FYI, you will need to rebase the changes made here, too https://github.com/awslabs/karpenter/pull/397

**Comment by clementrey-dev** [comment]
Sorry wrong manipulation with git

File: Makefile (hunk 1 above)
Comment (clementrey-dev) [reply Makefile]: agreed

File: Makefile (hunk 2 above)
Comment (clementrey-dev) [reply Makefile]: Nope, I was dynamic just to mimic the behaviour of RELEASE_MANIFEST variable. But RELEASE_MANIFEST had a dependency for CLOUD_PROVIDER (which is not the case of RELEASE_CHART).

Agreed

File: Makefile (hunk 3 above)
Comment (clementrey-dev) [reply Makefile]: Agreed

File: config/values.yaml (hunk 4 above)
Comment (clementrey-dev) [reply config/values.yaml]: Agreed

File: config/Chart.yaml (hunk 5 above)
Comment (clementrey-dev) [reply config/Chart.yaml]: Agreed

File: Makefile [hunk 6]
```diff
@@ -41,22 +40,22 @@ licenses: ## Verifies dependency licenses and requires GITHUB_TOKEN to be set
 	golicense hack/license-config.hcl karpenter
 
 apply: ## Deploy the controller into your ~/.kube/config cluster
-	$(WITH_GOFLAGS) ko apply -B -f config
+	helm template karpenter ./config/ | $(WITH_GOFLAGS) ko apply -B -f -
```
Comment (ellistarn) [inline Makefile:43]: nit: can you just do `./config` or better yet even `config`

File: Makefile [hunk 7]
```diff
@@ -41,22 +40,22 @@ licenses: ## Verifies dependency licenses and requires GITHUB_TOKEN to be set
...
 	./hack/codegen.sh
 
 publish: ## Generate release manifests and publish a versioned container image.
-	@aws ecr-public get-login-password --region us-east-1 | docker login --username AWS --password-stdin $(RELEASE_REPO)
-	$(WITH_RELEASE_REPO) $(WITH_GOFLAGS) ko resolve -B -t $(RELEASE_VERSION) --platform all -f config > $(RELEASE_MANIFEST)
-
-helm: ## Generate Helm Chart
-	cp $(RELEASE_MANIFEST) charts/karpenter/templates
+	@aws ecr get-login-password --region eu-west-1 | docker login --username AWS --password-stdin $(RELEASE_REPO)
+	@cp config/templates/* charts/karpenter/templates
```
Comment (ellistarn) [inline Makefile:53]: I'm a little confused here. Why does anything need to be copied? I thought the point was that all helm/config are now in the same location.

File: pkg/test/environment.go [hunk 8]
```diff
@@ -74,9 +74,9 @@ func NewEnvironment(options ...EnvironmentOption) *Environment {
 	ctx, stop := context.WithCancel(controllerruntime.SetupSignalHandler())
 	return &Environment{
 		Environment: envtest.Environment{
-			CRDDirectoryPaths: []string{project.RelativeToRoot("config")},
+			CRDDirectoryPaths: []string{project.RelativeToRoot("config/templates/provisioner-crd.yaml")},
 			WebhookInstallOptions: envtest.WebhookInstallOptions{
-				Paths: []string{project.RelativeToRoot("config")},
+				Paths: []string{project.RelativeToRoot("config/templates/webhooks.yaml")},
```
Comment (ellistarn) [inline pkg/test/environment.go:79]: You should be able to just pass /config/templates to both of these.

Comment (clementrey-dev) [reply pkg/test/environment.go:79]: If I just pass config/templates to `envtest`, then the test framework will not understand helm templating tag in controller.yaml and rbac.yaml and then failed.
Another way to work around would be before ginkgo something `helm template karpenter ./config/ > /.build/manfifest.yaml`
and then in file having somthing like :
   `CRDDirectoryPaths: []string{project.RelativeToRoot("/.build/manfifest.yaml")},`
   
 What would be best the project ?

File: Makefile (hunk 6 above)
Comment (clementrey-dev) [reply Makefile:43]: Agreed

File: Makefile [hunk 9]
```diff
@@ -41,22 +40,22 @@ licenses: ## Verifies dependency licenses and requires GITHUB_TOKEN to be set
...
+	$(WITH_RELEASE_REPO) $(WITH_GOFLAGS) ko resolve -B -t $(RELEASE_VERSION) --platform all -f config/values.yaml > charts/karpenter/values.yaml
 	yq e -i '.version = "$(RELEASE_VERSION)"' ./charts/karpenter/Chart.yaml
-	cd charts; helm package karpenter; helm repo index .
+
+helm:  ## Generate Helm Chart
+	cd charts; helm lint karpenter ;helm package karpenter; helm repo index .
```
[the remark this answers is not in the document]
Comment (clementrey-dev) [reply Makefile]: Fixed

File: Makefile (hunk 7 above)
Comment (clementrey-dev) [reply Makefile]: Sorry it was a misunderstanding, I moved all helm related in `charts/karpenter` and change the subsequent files (codegen and battletest)

File: pkg/test/environment.go (hunk 8 above)
Comment (ellistarn) [reply pkg/test/environment.go]: Gotcha sounds good.

**Review by ellistarn — APPROVED** [review]

**Comment by ellistarn** [comment]
Thank you for you work here!

**Pull request merged**